packages/data-api/src/validators.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { hasNoContent, takesDateForm, takesIdForm, yup, yupUtils } from '@tupaia/utils';
+
+export const eventOptionsValidator = yup.object().shape({
+  organisationUnitCodes: yup.array().of(yup.string().required()).strict().required(),
+  startDate: yup
+    .string()
+    .test(
+      yupUtils.yupTestAny([hasNoContent, takesDateForm], 'startDate should be in ISO 8601 format'),
+    ),
+  endDate: yup
+    .string()
+    .test(
+      yupUtils.yupTestAny([hasNoContent, takesDateForm], 'endDate should be in ISO 8601 format'),
+    ),
+  dataElementCodes: yup.array().of(yup.string().required()).strict(),
+  dataGroupCode: yup.string().strict().required(),
+  eventId: yup.string().test(yupUtils.yupTestAny([hasNoContent, takesIdForm])),
+});
+
+const aggregationStringValidator = yup.string().strict().required();
+const aggregationObjectValidator = yup.object().shape({
+  type: aggregationStringValidator,
+  config: yup.object(),
+});
+
+export const analyticsOptionsValidator = yup.object().shape({
+  organisationUnitCodes: yup.array().of(yup.string().required()).strict().required(),
+  startDate: yup
+    .string()
+    .test(
+      yupUtils.yupTestAny([hasNoContent, takesDateForm], 'startDate should be in ISO 8601 format'),
+    ),
+  endDate: yup
+    .string()
+    .test(
+      yupUtils.yupTestAny([hasNoContent, takesDateForm], 'endDate should be in ISO 8601 format'),
+    ),
+  dataElementCodes: yup.array().of(yup.string().required()).strict().required(),
+  aggregations: yup
+    .array()
+    .of<typeof aggregationStringValidator | typeof aggregationObjectValidator>(
+      yup.lazy((value: unknown) =>
+        typeof value === 'string' ? aggregationStringValidator : aggregationObjectValidator,
+      ) as any,
```

Comment:
This is a bit of a mind boggler

Reply:
Yeah I kinda hate it 😞 It's sort of a typescript error in `yup` where it can't properly infer the type of lazy validators when they're used in `yup.array().of()`. It's documented [here](https://github.com/jquense/yup/issues/1190), and seems like maybe it's been fixed in the current beta version?

Reply:
I'll add a comment and a TODO to upgrade yup once the fix goes out in a mainline release 👍